x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/ApiKey.java

```diff
@@ -36,8 +38,11 @@
     private final boolean invalidated;
     private final String username;
     private final String realm;
+    @Nullable
+    private final Map<String, Object> metadata;
```

Comment:
Why `Nullable` ?
Does a `null` metadata mean something different to empty?

Reply:
They should be functionally equivalent. The general approach I took is to let variables represent the underlying data exactly and only present them consistently when it is sent to end user. In this case, `ApiKey`'s metadata field can be `null` or empty map depending on the underlying data (the security index), where the metadata field can be `null`. When the API key information is returned to end user, the value of the metadata field will be consolidated to be `Map.of()` for both `null` and empty map (as shown in `ApiKey#toXContent`). That said, I am happy to change it to be an empty map, i.e. consolidation in its constructor.

Reply:
That's not our typical strategy. Usually we convert `null` to empty in the constructor so that objects that are semantically equivalent are also structurally equivalent.

Reply:
What you said makes sense. I was influenced by the existing `expiration` field which can be `null` when there is no expiration. But after some more thoughts, I see it is different from the `metadata` field because it does not have a meaningful default value other than `null`, while metadata has. I also agree that "objects that are semantically equivalent are also structurally equivalent" and that should be a rule to stick to. Thanks.